### PR TITLE
Total Variance Minimization with PyTorch support

### PR DESCRIPTION
**Pull request by MatteoFasulo**
# Description

Implemented Total Variation Minimization using PyTorch with GPU support, closely modeled after the existing CPU-based scipy.optimize.minimize approach. The CPU version iteratively minimizes over each image in the batch and each channel individually, resulting in time complexity of O(B × C) where B is the batch size and C the number of channels. Additionally, loss computation on the CPU introduces extra latency per iteration. The new PyTorch-based implementation enables parallelized, accelerated computation directly on the GPU, significantly improving runtime performance.

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Extensive test coverage was added via

```bash
pytest tests/defences/preprocessor/test_variance_minimization_pytorch.py --framework="pytorch"
```

- [x] test_variance_minimization_pytorch.py

validating:

- Memory efficiency
- Input dimension and shape handling
- Channel ordering (first vs. last)
- Output clipping within expected value ranges
- Convergence behavior
- Reproducibility across multiple runs with consistent seeds

**Test Configuration**:
- OS: Ubuntu 24.04 LTS
- Python version: Python 3.10.15
- ART version or commit number: 1.19.2
- TensorFlow / Keras / PyTorch / MXNet version: PyTorch

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My changes have been tested using both CPU and GPU devices

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @MatteoFasulo Thank you for your pull request, I'll add a review in the next few days.

File: tests/defences/preprocessor/test_variance_minimization_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import torch
+import numpy as np
+from numpy.testing import assert_array_equal
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_variance_minimization_pytorch.py:7]:
```suggestion
```

File: art/defences/preprocessor/variance_minimization_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+            # Place the optimized channel back into our result tensor
+            with torch.no_grad():
+                z_min[c, :, :] = res.view_as(z_min[c, :, :])
+
+        return z_min
+
+    @staticmethod
+    def _loss_func(
+        z_init: "torch.Tensor", x: "torch.Tensor", mask: "torch.Tensor", norm: float, lamb: float, eps=1e-6
```
Comment (beat-buesser) [inline art/defences/preprocessor/variance_minimization_pytorch.py:170]:
```suggestion
        z_init: "torch.Tensor", x: "torch.Tensor", mask: "torch.Tensor", norm: float, lamb: float, eps: float = 1e-6
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @MatteoFasulo Thank you very much for your pull request. I have added 3 minor change requests, could you please push the changes?

File: art/defences/preprocessor/variance_minimization_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+from typing import TYPE_CHECKING
+
+from tqdm.auto import tqdm
+
+import numpy as np
+
+from art.defences.preprocessor.preprocessor import PreprocessorPyTorch
+
+if TYPE_CHECKING:
+    from art.utils import CLIP_VALUES_TYPE
```
Comment (beat-buesser) [inline art/defences/preprocessor/variance_minimization_pytorch.py:23]:
```suggestion
if TYPE_CHECKING:
    import torch

    from art.utils import CLIP_VALUES_TYPE
```

**Comment by beat-buesser** [comment]
Hi @MatteoFasulo Thank you! I have missed this before, an additional import of torch is necessary for the typing to pass the Style Checks. Please add the code suggestion above and run Black formatter again to ensure correct line formatting. Afterwards we should be good to merge.

**Review by beat-buesser — APPROVED** [review]
Hi @MatteoFasulo Thank you very much for your contribution to ART!
